Use it Directly as Chromium extension

shotwn opened the pull request:
New script, capable of working as Chromium extension.

Currently github action updates latest release in every commit. This is probably unwanted, so small modification to github action is needed. (branch and on-commit lines)

A lot of commits to test github actions. So a squash might be needed also.

[New script](/chrome-extension/script.js) should still function as a TemperMonkey script, but I did not have time to test it. Hence I did not remove the main one on root.

Also adds README.MD so instructions are not based on each release.

shotwn committed:
feat: run as standalone chrome extension

shotwn committed:
fix: missing runs-on

shotwn committed:
fix: checkout extensions

shotwn committed:
test: release

shotwn committed:
test: upload artifact

shotwn committed:
test: upload artifact to release

shotwn committed:
feat: use non packed extensions

shotwn committed:
test: automatic chrome web store upload

shotwn committed:
chore: remove old github action

shotwn committed:
fix: action yml

shotwn committed:
fix: action yaml

shotwn committed:
test: actions

shotwn committed:
feat: make chrome extension store upload optional

shotwn committed:
feat: add permanent README.md

shotwn committed:
fix: extension action

shotwn committed:
chore: ignore webstore upload errors for now

shotwn committed:
feat: direct links in readme

shotwn committed:
chore: use direct down links instead of relative

okunamayanad commented:
Thank you so much for the contribition, I'll take consider this and probably merge the PR.

But first, I have a question, what did you meant by this?
> Currently github action updates latest release in every commit. This is probably unwanted, so small modification to github action is needed. (branch and on-commit lines)

shotwn commented:
Silly me, apparently, I already fixed it .
If you check _[chrome-extension.yml](https://github.com/okunamayanad/r-place-2023/pull/5/files#diff-7be551e27f23d1af5bb3ab2ebb7e566116b146c3dfe003cc7fb6b69f6d031f41)_:

1. It ZIPs the extension script and manifest.
2. Uploads artifact so it is available at github action result as a ZIP file.
3. Pushes chrome-extension.zip to the latest release's files.
4. Optionally uploads the extension to Chrome Web Store. Which needs a already published extension and Google API credentials.

It runs when a commit is pushed to chrome-extension branch or after any tag is created.
```
on:
  push:
    branches:
      - chrome-extension
    tags:
      - '**'
  ```

Step 3 should only run when a new tag created, my warning was about this. Because if it would change latest release's chrome-extension.zip on every commit push it wouldn't be logical. We want it to do step 3 only when we push a new release. Which I fixed with `if: ${{startsWith(github.ref, 'refs/tags/') }}` but forgot.

Meanwhile you still might want to change `branches: - chrome-extension` to `branches: - main` or your interim branch.

shotwn commented:
Test publish I pushed to the chrome webstore [passed review](https://chrome.google.com/webstore/detail/rplace-2023-tr-extension/lgafiioacndmgagmenljhcjckemnkide). 

Let me know if you want to release it yourself and I will remove it.

Kn0ax commented:
Uhm it seems really cool and thanks for the contribution but i think there is some issues with it

1. First of all Chrome Web Store have really slow update mechanizm and reviews tooks so long and if this event finish tonight or tomorrow it does not gonna make sense.

2.Most of the users are not comfortable with installing chrome extentions and the reason they mostly agrees is they are able to see script code & updates and with extentions it does not clear from user side.

3.I might be wrong but chrome is not usually okey with getting executed code from internet so it means when we need to update we have to publish a update and it triggers again first issue. 

What do you think about it @okunamayanad ?

shotwn commented:
1. My web store version took 5 hours to publish and updates to the script take around an hour. But here is the thing, extension does not need to be updated to update overlay image. It is not embedded to the extension, it still uses image from github repo, so it is pretty much momentary. Just like the original script, which was not updated for a while when overlays do. 
2. Extension is vetted by google and has thight permissions but of course anyone is free to see the code. This is why chrome web store upload it only optional. If you check readme.md of the fork, primary method of using the script as extension is to just download extension zip, unzip it, and load it in to chrome in dev mode. Script is open and readable in zipped extension folder alongside manifest. We still use the script openly but we take away installing temper monkey part away.
3. Extension has no problem using the overlay picture from github. So overlay image can be updated to all your heart desire and will be effective immediately. Script which loads the picture gets updated very rarely anyway. But when it does, manual installation methods need user interaction while web store version is automatic. That is the only difference.

So there is no difference between temper monkey install and direct unpacked extension install. Meanwhile chrome web store install gives google's vetting and auto-update. 

Overlay image will get updated live in any version. Script update has to be done manually in manual installations meanwhile webstore will update it automatically (after 1 hour or so, google scans packages).

Also tech savy users can still observe source code of web store version, in their local appdata/chrome folder, source is extracted just like manual install, and .js file is readable.

Even as tempermonkey script new version has better sanity checks and debug logging, so it is an improvement.

shotwn commented:
For after the fork merge, I was also planning to add overlay image ping, so system would refresh the image without need of page refresh and a button to turn the overlay off/on.

But Monday is upon us and one has to work, so I probably have no time left.

shotwn committed:
chore: new output image src

shotwn committed:
bump version

shotwn committed:
bump version

shotwn committed:
Merge branch 'okunamayanad:main' into chrome-extension

okunamayanad commented:
Didn't merge it 'cause it'd cause a lot of misunderstanding, closing the pr now because r/place has ended

Although, I'd like to see this for the next r/place, contact me @ discord: okunamayanad or email: baran@okunamayanad.com
